### PR TITLE
Move Compiler.Service build and test to FSharp

### PR DESCRIPTION
**Pull request by KevinRansom**
The fcs branch is very unreliable, so moving it to our regular build.

Still need to work out some stuff, esp.  version numbers.

### CODE REVIEW

**Comment by auduchinok** [comment]
Just in case: please keep the existing FSharp.Compiler.Solution solution around too. It's very helpful to have FSharp.Core as an external dependency (since it's how tools like Ionide or Rider get it).

**Comment by auduchinok** [comment]
@cartermp No, I meant when working on FSharp.Compiler.Service solution, it's much easier when FSharp.Core comes as a NuGet package and not as a project reference.

**Comment by cartermp** [comment]
How is it easier? This wasn't something discussed in the leadup to the RFC: https://github.com/fsharp/fslang-design/blob/master/tooling/FST-1008-publish-fcs-from-upstream-repo.md

The main quirk today is that FCS still targets an older FSharp.Core for reasons that feel ... strange. We generally feel it should be like the compiler and build/ship in lockstep with FSharp.Core. I know that a while back @dsyme disagreed because there may be people using FCS in older environments. But with an eventual move to .NET Standard 2.0 that sort of feels less important and more to the point, an unlikely scenario (brand-new syntax in an older environment) that isn't worth trying to support,

**Comment by auduchinok** [comment]
> How is it easier?

@cartermp Sorry for the long reply. There are several reasons:
* FCS clients use it as a library and it's important to align all dependencies (including FSharp.Core) properly in the resulting application (including libraries like Fantomas which in turn used as a library; including people who host FCS in their games or apps other than code editing tools)
* Having FSharp.Core as a part of the solution implies it needs to be type checked too (so working on FCS becomes slower and uses more memory) while there's no need for it when working on FCS (as opposed to FCP in the actual compiler).
* Having FSharp.Core in a solution enables some special analysis execution paths and workarounds, and it's not always supported well in editor tools (VS and Ionide had issues with it several times in the past, Rider has a few issues now), so unless you actually need working on FSharp.Core (the FCT scenario) it's better to see it as a referenced assembly, ensuring everything works smoothly even if FSharp.Core workarounds get broken in a particular tool.

I'd be happy with FSharp.Core version in FCS matching the one SDK implicitly references, so it's fresh enough but still is referenced as a package. Since we've been concerned about greater reach in the past, an LTS SDK could work well, though, I'm not sure if can be too outdated for realistic FCS use scenarios.

I'm also happy with the switch @KevinRansom has added if the resolution is to go with a project reference by default. It'd probably good to have a test run with the switch enabled though.

**Comment by dsyme** [comment]
I haven't thought about this for a while, but I guess the basic issue is still that FCS is at the root of an ecosystem (e.g. FSharp.Formatting, or whatever), and if FCS depends on FSharp.Core latest then that entire ecosystem depends on FSharp.Core latest.

I guess the equivalent would be Roslyn depending on very very recent .NET libraries so only people on .NET CoreApp 3.1 or something (rather than ,NET standard 2.1 or whatever the dependency is) could use latest Roslyn.  

I don't know how much that matters at the moment.  It feels architecturally wrong for such a fundamental component to induce a latest-dependency (but equally most consuming tools will probably be ok with that on package update)

Of course it would be wrong for FCS builds to depend on unpublished FSharp.Core, that goes without saying doesn't it?  That would render those components useless until the corresponding FSharp.Core is on nuget.

**Comment by dsyme** [comment]
BTW why don't we do the following?

* Build both FSharp.Compiler.Private and FSHarp.Compiler.Service against the same, agreed, stable, published FSharp.Core (but not the one we build in-repo), so the dependency is say on 4.7.2.  

* Build apps (Fsi.exe, fsc.exe) and tests (FSharp.*.Tests.dll) against the one we build in-repo 

Basically there's no need to build FSharp.Compiler.Private against the latest FSharp.Core si there?  That would solve all our problems wouldn't it?

**Comment by abelbraaksma** [comment]
> Basically there's no need to build FSharp.Compiler.Private against the latest

You may lose certain recent optimizations though like in arrays and strings, which are only available in FSharp.Core. 

In the case of the recent interpolated strings feature, that depends on a future FSharp.Core. If you would wanted to use that feature inside FSharp.Compiler.Private, you can't make that dependency change.

But I agree that it feels much better if compiler and FCS can be independent and rely on an older (not latest) version. The benefits probably outweigh the costs, and in forthcoming scenarios, we may consider a point release for FSharp.Core to prevent the divide for getting too wide. 

Just my 2p ;)

**Comment by dsyme** [comment]
> In the case of the recent interpolated strings feature, that depends on a future FSharp.Core...

Yup (note we already have these restrictions because the FCS build has to build the same code using a published FSharp.Core)

File: tests/service/data/CSharp_Analysis/CSharp_Analysis.csproj [hunk 1]
```diff
@@ -2,7 +2,7 @@
 <Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFramework>netstandard2.0</TargetFramework>
+    <TargetFramework>netstandard20</TargetFramework>
```
Comment (dsyme) [inline tests/service/data/CSharp_Analysis/CSharp_Analysis.csproj:5]: Good catch

File: tests/service/ExprTests.fs [hunk 2]
```diff
@@ -991,6 +991,7 @@ let testOperators dnName fsName excludedTests expectedUnoptimized expectedOptimi
     let projFilePath = Utils.getTempFilePathChangeExt tempFileName ".fsproj"
 
     try
+        System.Diagnostics.Debugger.Break()
```
Comment (dsyme) [inline tests/service/ExprTests.fs:994]: What's this?

File: tests/service/Common.fs [hunk 3]
```diff
@@ -118,6 +120,7 @@ let mkProjectCommandLineArgsSilent (dllName, fileNames) =
         yield "--target:library" 
         for x in fileNames do 
             yield x
+        System.Diagnostics.Debugger.Break()
```
Comment (dsyme) [inline tests/service/Common.fs:123]: What's this?

File: tests/service/Common.fs [hunk 4]
```diff
@@ -14,6 +15,7 @@ open NUnit.Framework
 #if NETCOREAPP
 let readRefs (folder : string) (projectFile: string) =
     let runProcess (workingDir: string) (exePath: string) (args: string) =
+        System.Diagnostics.Debugger.Break()
```
Comment (dsyme) [inline tests/service/Common.fs:18]: Another one

File: src/fsharp/FSharp.Compiler.Service/FSharp.Compiler.Service.fsproj [hunk 5]
```diff
@@ -0,0 +1,778 @@
...
+
+  <PropertyGroup>
+    <!--
+    Some users have found it easier to work on FSharp.CompilerService when referencing the nuget package rather than the fsharp.core project
+    set FSHARPCORE_USE_PACKAGE to true to use package
+  -->
+    <FSHARPCORE_USE_PACKAGE>false</FSHARPCORE_USE_PACKAGE>
+  </PropertyGroup>
+
+  <ItemGroup Condition="'$(FSHARPCORE_USE_PACKAGE)' != 'true'">
```
Comment (dsyme) [inline src/fsharp/FSharp.Compiler.Service/FSharp.Compiler.Service.fsproj:736]: As discussed in comments, I think it's best to change both FSharp.Compiler.Private and FSharp.Compiler.Service to compile against a stable published FSharp.Core and drop their refrence to FSharp.Core.fsproj.  At least FSharp.Compiler.Service should continue to be compiled this way.

**Comment by cartermp** [comment]
> I haven't thought about this for a while, but I guess the basic issue is still that FCS is at the root of an ecosystem (e.g. FSharp.Formatting, or whatever), and if FCS depends on FSharp.Core latest then that entire ecosystem depends on FSharp.Core latest.

FSharp.Formatting requires at least FSharp.Core 4.7 as per its dependency tree. I suspect nearly all components that _use_ FCS rather than depend on it for generating docs are similarly depending on a higher FSharp.Core. I've seen little evidence that FCS is at the root of an ecosystem that cannot consume the latest FSharp.Core.

> I guess the equivalent would be Roslyn depending on very very recent .NET libraries so only people on .NET CoreApp 3.1 or something (rather than ,NET standard 2.1 or whatever the dependency is) could use latest Roslyn.

Not quite the same, but the code analysis stuff is moving to NS 2.0-only and therefore locking out a good lot of the older C# ecosystem. If VS didn't require components to be compatible with .NET Framework then it's reasonable to believe they would be targeting .NET Core. Though they have a fundamentally different architecture and aren't so tied to a library like we are. 

The main issue here is that so much of F# is intrinsically tied to FSharp.Core that the two are completely inseparable. FSharp.Core is not just some F# library, it literally contains half the implementation for numerous features. If we want to purposefully tie anything in FCS to an older F# language version, then we cannot use anything good we've built inside our own codebase, nor allow any distributed compiler for tools benefit from their use either.

> Of course it would be wrong for FCS builds to depend on unpublished FSharp.Core, that goes without saying doesn't it? That would render those components useless until the corresponding FSharp.Core is on nuget.

Not really, since the built FSharp.Core is available via a myget feed. Any new FCS behavior is subject to change as we evolve the language and that has resulted in many breaking changes to behavior for downstream consumers just in this past year. I don't see that changing as we iterate on designs for things. This isn't typically related to FSharp.Core, but if the goal is compatibility then I think we've failed pretty hard on that front.

> Build both FSharp.Compiler.Private and FSHarp.Compiler.Service against the same, agreed, stable, published FSharp.Core (but not the one we build in-repo), so the dependency is say on 4.7.2.

Doesn't that lock us out of future developments on numerous language features? I don't see how we could have done the Witness Passing feature without that, for example. Not without some absolutely atrocious hacks in place.

I think it's not a good goal to have newer versions of FCS rely on older versions of the F# core library. I've only ever seen that cause problems, such requiring us to do things like build our own `ValueTuple` even though the same type with more module-bound functionality is just sitting there for us. This is needless complexity and even worse for contributors to deal with.

**Comment by abelbraaksma** [comment]
Just a shot in the dark, but what if we statically link FSharp.Core? That way, the runtime dependency disappears and we only have the compile time dependency, and can use whatever version, including the source version, we want.

(I don't know if this would open another can of worms, though)

**Comment by KevinRansom** [comment]
@abelbraaksma ,  static linking FSharp.Core is not going to be very useful.  We need a bunch of separately compiled assemblies to have a single understanding of fsharp core types for example List.  If any of those assemblies had an embedded FSharp.Core then they would disagree about those typesand couldn't share them.

@dsyme ,
If FSharp.Core was a library assembly, then sure we could take a dependency on it.  However, a lot of our compiler features are implemented with their support in FSharp.Core.  For example, how much of interpolated strings would work if FCS was pulling the FSharp.Core from 4.6.2 or even 4.7.2.   That's right, very little.

When a language feature developer I.e. you ... wants to build a new feature, they should consider how it surfaces in FCS and as such would need to build, deploy and test with the latest FSharp.Core.n Failing that, they would have to wait until we ship an RTM FSharp.Core with the language feature accepted, and then work on improving FCS ... or ... hack up the build themselves to load the project, of course then the tests wouldn't work so they would need to hack them up too.

All things considered building language features with a hard dependency on FSharp.Core is probably a bad design choice,  however, it is our goto implementation method for nearly every language feature, and so it is what we have got.

We already have had to figure out how to deploy preview FSharp.Core versions, they are dropped to the same feed as the rest of .NET pushes preview nuget packages.  And they are dropped at least every day, that there is a change in the repo.

We would publish preview FSharp.Compiler.Service packages to that every day.  So if a developer wanted to be bang up to date, then she would just grab the FCS and FSharp.Core from there.  And then if you folks wanted we would publish to nuget an Fsharp.Compiler.Service, every RTM release of the fsharp.compiler, which corresponds to a specific .NetSDK.

The reason that is a good plan, is that all of the tooling and the compilers would all be aligned and built from the same codebase at the same time.  Anyone who wanted a specific improvement can grab it early from the preview channel.

However, if you want to continue to own the build and deployment of FCS that is fine too, however, we will still use this project in order to build and run the FCS tests, because the fcs folder off of this repo is too unreliable.


Kevin

File: tests/service/ExprTests.fs (hunk 2 above)
Comment (KevinRansom) [reply tests/service/ExprTests.fs:994]: carelessly left over rubbish from debugging.

File: tests/service/Common.fs (hunk 3 above)
Comment (KevinRansom) [reply tests/service/Common.fs:123]: carelessly left over rubbish from debugging.

File: tests/service/Common.fs (hunk 4 above)
Comment (KevinRansom) [reply tests/service/Common.fs:18]: carelessly left over rubbish from debugging.

File: tests/service/data/CSharp_Analysis/CSharp_Analysis.csproj (hunk 1 above)
Comment (brettfo) [reply tests/service/data/CSharp_Analysis/CSharp_Analysis.csproj:5]: `NETSTANDARD20` is the compiler constant introduced by MSBuild; `netstandard2.0` is still the correct TFM.

File: tests/FSharp.Compiler.Service.Tests/FSharp.Compiler.Service.Tests.fsproj [hunk 6]
```diff
@@ -0,0 +1,110 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFrameworks>net472;netcoreapp3.1</TargetFrameworks>
+    <TargetExt>.dll</TargetExt>
+    <RuntimeIdentifiers>win-x86;win-x64;linux-x64;osx-x64</RuntimeIdentifiers>
+    <TargetFrameworks Condition="'$(OS)' == 'Unix'">netcoreapp3.1</TargetFrameworks>
```
Comment (brettfo) [inline tests/FSharp.Compiler.Service.Tests/FSharp.Compiler.Service.Tests.fsproj:8]: Nit: can you move this line up next to the original `<TargetFrameworks>` line?  It's easier to find.

File: tests/FSharp.Compiler.Service.Tests/CSharp_Analysis/CSharp_Analysis.csproj [hunk 7]
```diff
@@ -0,0 +1,10 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFrameworks>netstandard20</TargetFrameworks>
```
Comment (brettfo) [inline tests/FSharp.Compiler.Service.Tests/CSharp_Analysis/CSharp_Analysis.csproj:3]: `netstandard2.0` is the appropriate TFM.

File: tests/service/data/CSharp_Analysis/CSharp_Analysis.csproj (hunk 1 above)
Comment (cartermp) [inline tests/service/data/CSharp_Analysis/CSharp_Analysis.csproj:5]: this still needs to be `netstandard2.0`

**Pull request merged**